<a href="https://colab.research.google.com/github/ShinAsakawa/xerion/blob/master/notebooks/xerion_demo001.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Disclaimer of Xerion, data management for simulating SM89 and PMSP96

for [wbai handson](https://wba-initiative.org/3411/)

<div>
    <center><img src="https://wba-initiative.org/wp-content/uploads/2015/05/logo.png" style="width:29%"></center>
</div>
    

<p style='text-align:center;'>
    <font color='green' size='+1' style='font-weight:bold;'>Shin Asakawa &lt;asakawa@ieee.org&gt;</font>
</p>

---
# 手順1. `Git clone` する

In [1]:
# GitHub から必要なパッケージを入手してください
!git clone https://github.com/ShinAsakawa/xerion.git

fatal: destination path 'xerion' already exists and is not an empty directory.


In [1]:
# numpy の不都合によりバージョンを変更します
!pip install --user -U numpy==1.16.0

Requirement already up-to-date: numpy==1.16.0 in /root/.local/lib/python3.6/site-packages (1.16.0)


In [2]:
# 上で入手したパッケージをインストールします
!pip install --user -U setuptools
!pip install --user --upgrade ./xerion

Requirement already up-to-date: setuptools in /usr/local/lib/python3.6/dist-packages (40.8.0)
Processing ./xerion
  Stored in directory: /tmp/pip-ephem-wheel-cache-v979hs3n/wheels/4b/97/24/8adbd8d8f37b7470911c06e8466e900919e2e8f99a6c9e656a
Successfully built xerion
  Found existing installation: xerion 0.3
    Uninstalling xerion-0.3:
      Successfully uninstalled xerion-0.3


In [5]:
# インストール結果を確認します
import numpy as np
from xerion.xerion import Xerion

data = Xerion()
##descr() はデータの諸元を表示します
#data.descr()

self.datadir=/content/xerion/xerion/data/
self.datafilename=/content/xerion/xerion/data/SM-nsyl.pkl


---
# 手順2. データの読み込み

In [6]:
import numpy as np
from xerion.xerion import Xerion
handson = Xerion()

# Xerion は data 管理用モジュールです。以下のコメントを外してご覧ください
#handson.descr()
#handson.usage()

self.datadir=/content/xerion/xerion/data/
self.datafilename=/content/xerion/xerion/data/SM-nsyl.pkl


In [7]:
grapheme = handson.db['grapheme']
phoneme = handson.db['phoneme']
seq = handson.db['seq']
len(handson.db['grapheme'])

2998

---

# 手順3. 読み込んだデータの確認

ここに `input`, `output`, `freq`, `grapheme`, `phoneme`, `seq`, `tag` でアクセスできるようにしてあります。

- input: np.ndarrray((2998, 105), dtype=float32) # 文字単語のトリプレット表現 105 次元ベクトル
- output: np.ndarray((2998, 61), dtype=float32)  # 対応する音韻トリプレット表現 61 次元ベクトル
- frep: np.ndarray((2998,), dtype=float32)       # 対応する頻度情報
- seq: 
- grapheme: 入力文字列リスト
- phoneme: 出力音韻列リスト ARPABET 表現
- tag: 無視してください

SM89, PMSP96 では意味層は実装されていません。<br>
以下に例を示します

In [8]:
print(handson.tags)
handson.grapheme, handson.phoneme, handson.freq, handson.seq, handson.inputs, handson.outputs, set(handson.tag)

('#', 'seq', 'grapheme', 'phoneme', 'freq', 'tag', 'inputs', 'outputs')


(array(['ace', 'ache', 'act', ..., 'zone', 'zoo', 'zounds'], dtype='<U8'),
 array(['As', 'Ak', '@kt', ..., 'zOn', 'zU', 'zWnz'], dtype='<U6'),
 array([0.236101, 0.149313, 0.471041, ..., 0.213746, 0.199825, 0.115525],
       dtype=float32),
 array([   0,    1,    2, ..., 2995, 2996, 2997], dtype=int16),
 array([[0, 0, 0, ..., 1, 0, 0],
        [0, 0, 0, ..., 1, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 1, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=int16),
 array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=int16),
 {'#',
  'AMB',
  'CON',
  'EXC',
  'EXPT',
  'HEC',
  'HFE',
  'HFEEXPT',
  'HFRIC',
  'HRI',
  'HSTR',
  'HSTRUNQ',
  'LEC',
  'LFE',
  'LFEEXPT',
  'LFRI',
  'LRIC',
  'LRICCON',
  'LSTR',
  'LSTRUNQ',
  'REG',
  'RI',
  'STR',
  'STREXPT',


In [0]:
##最初，このセルは無視してください
#print(handson.url_base, handson.url_file)
#print(handson.basefilename)
#print(handson.syl_files, '\n', handson.nsyl_files)
#print(handson.datadir)
#print(handson.datafilenames)
#print(type(handson.db), len(handson.db))

In [10]:
type(handson.db)
for k in handson.db.keys():
    print(k, end=' ')
    if isinstance(handson.db[k], list):
        print(len(handson.db[k]))
    elif isinstance(handson.db[k], np.ndarray):
        print(handson.db[k].shape)

inputs (2998, 105)
outputs (2998, 61)
grapheme (2998,)
phoneme (2998,)
freq (2998,)
tag 2998
seq (2998,)


In [11]:
## 各データを表示します
handson.freq        ### データ内の各単語の頻度
handson.grapheme    ### 各単語の表記表現
handson.phoneme     ### 各単語の音韻表現
len(handson.inputs), handson.inputs.shape, handson.inputs     ###入力表現 numpy の ndarray
len(handson.outputs), handson.outputs.shape, handson.outputs  ###出力表現 numpy の ndarray

(2998, (2998, 61), array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=int16))

In [12]:
#以下に 3 連表現を示します。PMSP96論文中では triplet と呼ばれていたオンセット onset，母音 vowel, コーダ coda です。
for item in [handson.Orthography, handson.Phonology]:
    for entry in ['onset', 'vowel', 'coda']:
        print(item[entry])

['Y', 'S', 'P', 'T', 'K', 'Q', 'C', 'B', 'D', 'G', 'F', 'V', 'J', 'Z', 'L', 'M', 'N', 'R', 'W', 'H', 'CH', 'GH', 'GN', 'PH', 'PS', 'RH', 'SH', 'TH', 'TS', 'WH']
['E', 'I', 'O', 'U', 'A', 'Y', 'AI', 'AU', 'AW', 'AY', 'EA', 'EE', 'EI', 'EU', 'EW', 'EY', 'IE', 'OA', 'OE', 'OI', 'OO', 'OU', 'OW', 'OY', 'UE', 'UI', 'UY']
['H', 'R', 'L', 'M', 'N', 'B', 'D', 'G', 'C', 'X', 'F', 'V', '∫', 'S', 'Z', 'P', 'T', 'K', 'Q', 'BB', 'CH', 'CK', 'DD', 'DG', 'FF', 'GG', 'GH', 'GN', 'KS', 'LL', 'NG', 'NN', 'PH', 'PP', 'PS', 'RR', 'SH', 'SL', 'SS', 'TCH', 'TH', 'TS', 'TT', 'ZZ', 'U', 'E', 'ES', 'ED']
['s', 'S', 'C', 'z', 'Z', 'j', 'f', 'v', 'T', 'D', 'p', 'b', 't', 'd', 'k', 'g', 'm', 'n', 'h', 'I', 'r', 'w', 'y']
['a', 'e', 'i', 'o', 'u', '@', '^', 'A', 'E', 'I', 'O', 'U', 'W', 'Y']
['r', 'I', 'm', 'n', 'N', 'b', 'g', 'd', 'ps', 'ks', 'ts', 's', 'z', 'f', 'v', 'p', 'k', 't', 'S', 'Z', 'T', 'D', 'C', 'j']


## 同形異音語について

In [13]:
len(handson.grapheme), type(handson.grapheme)
grapheme_set = set(handson.grapheme)
len(grapheme_set)

print('データ総数:{0} とユニークなデータ数:{1} との差が同形異音語数になります'.format(
    len(handson.grapheme),len(grapheme_set)))

データ総数:2998 とユニークなデータ数:2985 との差が同形異音語数になります


In [14]:
# 実際に同形異音語を表示してみましょう
prev_word = ""
n_homographs = 0
for word in sorted(handson.grapheme):
    if word == prev_word:
        n_homographs += 1
        print(word, end=' ')
    prev_word = word
print('\nn_homographs={}'.format(n_homographs)) 

bass bow dove house lead live read route row sow tear wind wound 
n_homographs=13


In [0]:
### 以下は知らなくても良い情報ですので無視してください for debuging
#handson.load_pickle()
#data.make_all()
#print(handson.module_path)
#handson.note()
#handson.nsyl_files
#handson.origfile_size
#print(handson.Orthography)
#print(handson.Phonology)
#print(handson.pkl_dir)
#handson.url_base
#handson.url_file
#handson.xerion_prefix

In [16]:
for db in handson.dbs:
    print('database: ', db) # , db.dbs[db][3].shape, type(db.dbs[db][3])) #, x.dbs[db][3])

database:  SM-nsyl


---
# データないに登録された各単語の発音を調べる

- pyhon で自然言語処理する際に使われる `nltk` を使って単語の発音を調べてみます。
- `nltk` については [https://www.nltk.org/](https://www.nltk.org/) をご覧ください

In [17]:
# Colaboratory では cmudict がダウンロードされいないようなのでダウンロードします
# 一度だけ実行すればよいので，その都度実行する必要はありません
import nltk
nltk.download('cmudict')

[nltk_data] Downloading package cmudict to /root/nltk_data...
[nltk_data]   Unzipping corpora/cmudict.zip.


True

In [18]:
from nltk.corpus import cmudict
arpabet = cmudict.dict()

# 1. Xerion の単語を取り出します
words = handson.db['grapheme']

# 2. Xerion に存在する単語のうち cmudict に発音の登録のない単語を抜き出します。
not_in_arpabet = [word for word in words if not word in arpabet]

print('{0}/{1}={2:.2f} percent words in Xerion data were not registered in ARPABET.'
      .format(len(not_in_arpabet), len(words), len(not_in_arpabet)/len(words)*100))

# 随分ありますね。
print(not_in_arpabet)

55/2998=1.83 percent words in Xerion data were not registered in ARPABET.
['awn', 'bas', 'bilge', 'bleat', 'calve', 'cheep', 'chive', 'clod', 'clop', 'clove', 'crag', 'cud', 'dork', 'fiche', 'flays', 'frappe', 'git', 'gunk', 'hasp', 'jag', 'jape', 'laze', 'leer', 'letch', 'manse', 'muss', 'nape', 'paunch', 'phage', 'pram', 'rend', 'retch', 'rheum', 'rime', 'runt', 'scrim', 'shoal', 'shush', 'snoot', 'sooth', 'souse', 'sprig', 'squaw', 'thwack', 'tog', 'tyme', 'ump', 'veldt', 'vend', 'whelp', 'whil', 'whir', 'wretch', 'yawl', 'zit']


In [19]:
# ちなみに cmu 辞書についてですが
print(len(arpabet))

arpabet_vocab = [v for v in arpabet.keys()]
print(arpabet_vocab[:5])
arpabet_sounds = [s for s in arpabet.values()]
print(arpabet_sounds[:5])
#上記の表記については ARPABET をご覧ください

123455
['a', 'a.', 'a42128', 'aaa', 'aaberg']
[[['AH0'], ['EY1']], [['EY1']], [['EY1', 'F', 'AO1', 'R', 'T', 'UW1', 'W', 'AH1', 'N', 'T', 'UW1', 'EY1', 'T']], [['T', 'R', 'IH2', 'P', 'AH0', 'L', 'EY1']], [['AA1', 'B', 'ER0', 'G']]]


---
# ARPABET に登録されていない単語の処理

In [20]:
from itertools import product as iterprod
#arpabet = nltk.corpus.cmudict.dict()

def wordbreak(s):
    """
    See https://stackoverflow.com/questions/33666557/get-phonemes-from-any-word-in-python-nltk-or-other-modules

    """
    s = s.lower()
    if s in arpabet:
        return arpabet[s]
    middle = len(s)/2
    partition = sorted(list(range(len(s))), key=lambda x: (x-middle)**2-x)
    for i in partition:
        pre, suf = (s[:i], s[i:])
        if pre in arpabet and wordbreak(suf) is not None:
            return [x+y for x,y in iterprod(arpabet[pre], wordbreak(suf))]
    return None

# nltk の cmudict に存在しない単語の読みを表示
for word in not_in_arpabet:
    print(word, wordbreak(word))

awn [['AO1', 'EH1', 'N']]
bas [['B', 'IY1', 'AE1', 'Z'], ['B', 'IY1', 'EH1', 'Z']]
bilge [['B', 'IH1', 'L', 'JH', 'IY1', 'IY1']]
bleat [['B', 'L', 'IY1', 'T', 'IY1']]
calve [['K', 'AE1', 'L', 'V', 'IY1'], ['K', 'AE1', 'L', 'V', 'IY1', 'IY1']]
cheep [['CH', 'EY1', 'IY1', 'P', 'IY1']]
chive [['K', 'AY1', 'V', 'IY1'], ['K', 'AY1', 'V', 'IY1', 'IY1']]
clod [['S', 'IY1', 'L', 'OW1', 'D', 'IY1']]
clop [['S', 'IY1', 'L', 'AA1', 'P']]
clove [['S', 'IY1', 'L', 'AH1', 'V']]
crag [['S', 'IY1', 'R', 'AE1', 'G']]
cud [['S', 'IY1', 'Y', 'UW1', 'D', 'IY1']]
dork [['D', 'UW1', 'AA1', 'R', 'K', 'EY1']]
fiche [['F', 'AY1', 'CH', 'EY1'], ['F', 'IY1', 'CH', 'EY1']]
flays [['F', 'L', 'EY1', 'EH1', 'S']]
frappe [['F', 'R', 'AE1', 'P', 'P', 'IY1', 'IY1']]
git [['JH', 'IY1', 'IH1', 'T'], ['JH', 'IY1', 'IH0', 'T']]
gunk [['G', 'UW1', 'EH1', 'N', 'K', 'EY1']]
hasp [['HH', 'AA1', 'EH1', 'S', 'P', 'IY1'], ['EY1', 'CH', 'EY1', 'EH1', 'S', 'P', 'IY1']]
jag [['Y', 'AA1', 'JH', 'IY1']]
jape [['Y', 'AA1', 'P', 'IY1', 

---
# 意味層の実装

from <https://fasttext.cc/docs/en/pretrained-vectors.html>

In [0]:
# fastText データの読み込み。次の行頭の # を外してください
#!time wget https://dl.fbaipublicfiles.com/fasttext/vectors-wiki/wiki.en.vec

--2019-03-18 00:13:06--  https://dl.fbaipublicfiles.com/fasttext/vectors-wiki/wiki.en.vec
Resolving dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)... 104.20.6.166, 104.20.22.166, 2606:4700:10::6814:16a6, ...
Connecting to dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)|104.20.6.166|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6597238061 (6.1G) [binary/octet-stream]
Saving to: ‘wiki.en.vec’

wiki.en.vec          70%[=============>      ]   4.34G  38.4MB/s    eta 78s    

In [0]:
word2vec_file = 'wiki.en.vec'
with open(word2vec_file, 'r') as f:
    lines = f.readlines()

(list, 2998)

In [0]:
words_list = list(handson.grapheme)
type(words_list), len(words_list)
semantics_wiki = {}

(list, 2998)

In [0]:
%%timeit
for line in lines:    
    buff = line.strip().split(' ')
    word = buff[0]
    if word in words_list:
        semantics_wiki[word] = np.asarray([float(x) for x in buff[1:]],dtype=np.float32)

#同形異音語があるので，その数だけ semantics は数が少ないです。
len(semantics_wiki), type(semantics_wiki)
semantics = np.ndarray((len(words_list), 300), dtype=np.float32)
# そこで数合わせをします

2min 15s ± 809 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [0]:
semantics = np.ndarray((len(words_list), 300), dtype=np.float32)

In [0]:
%%timeit
for i, w in enumerate(words_list):
    semantics[i] = semantics_wiki[w]

2.48 ms ± 118 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [0]:
handson.sememe = semantics
for modality in [handson.grapheme, handson.phoneme, handson.sememe]:
    len(modality), type(modality), modality.shape

(2998, numpy.ndarray, (2998,))

(2998, numpy.ndarray, (2998,))

(2998, numpy.ndarray, (2998, 300))

## 視覚化

In [0]:
from sklearn.decomposition import PCA

G = handson.db['inputs']
pca = PCA(n_components=4)
pca.fit(G)
print(pca.explained_variance_ratio_) 

S = handson.sememe
pca.fit(S)
print(pca.explained_variance_ratio_) 

P = handson.db['outputs']
pca.fit(P)
print(pca.explained_variance_ratio_)

PCA(copy=True, iterated_power='auto', n_components=4, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

[0.07402131 0.058128   0.05556856 0.04860908]


PCA(copy=True, iterated_power='auto', n_components=4, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

[0.02843864 0.02434086 0.02053765 0.01836787]


PCA(copy=True, iterated_power='auto', n_components=4, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

[0.05501097 0.04969663 0.0479385  0.0450488 ]


In [0]:
pca.get_params()

{'copy': True,
 'iterated_power': 'auto',
 'n_components': 2,
 'random_state': None,
 'svd_solver': 'auto',
 'tol': 0.0,
 'whiten': False}

In [0]:
handson.input.shape
len(semantics)
handson.grapheme[:3]

In [0]:
for word in handson.grapheme[:3]:
    print(word, semantics[word].shape, semantics[word].dtype)

In [0]:
#type(semantics)
type(handson.grapheme), handson.grapheme.shape, handson.grapheme[:3]
type(handson.phoneme), handson.phoneme.shape, handson.phoneme[:3]
type(handson.sememe), handson.sememe.shape, handson.sememe[:3]
type(handson.freq), handson.freq.shape, handson.freq[:3]

---

# 同形異音文字の表示
<!--
- [](https://en.wikipedia.org/wiki/File:Homograph_homophone_venn_diagram.svg)
- [](https://upload.wikimedia.org/wikipedia/commons/thumb/f/fa/Homograph_homophone_venn_diagram.svg/1280px-Homograph_homophone_venn_diagram.svg.png)
-->

<div>
<center>
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/fa/Homograph_homophone_venn_diagram.svg/1280px-Homograph_homophone_venn_diagram.svg.png" style="width:29%"><br>
from [https://en.wikipedia.org/wiki/Heteronym_(linguistics)](https://en.wikipedia.org/wiki/Heteronym_(linguistics))<br>

from <https://en.wikipedia.org/wiki/Heteronym_(linguistics)>
</center>

see also <http://www.singularis.ltd.uk/bifroest/misc/homophones-list.html>
</div>

In [0]:
words = handson.graph # orthography
prev_word = None
same_count = 0
for i, word in enumerate(words):
    if not word in semantics:
        print(i, word)  #, wrd2idx[word], semantics[word])
        del orthography[word]
        del phonology[word]
    if prev_word == word:
        same_count += 1
        print(word, end=' ')
    prev_word = word